toml/src/main/kotlin/org/rust/toml/resolve/CargoTomlFileReferenceProvider.kt

```diff
@@ -39,11 +42,20 @@ class CargoTomlFileReferenceProvider(private val patternType: PathPatternType) :
         }
 
         if (kind !is TomlLiteralKind.String) return emptyArray()
-        return CargoTomlFileReferenceSet(element, completeDirs, completeRustFiles).allReferences
+        val referenceSet = CargoTomlFileReferenceSet(element, completeDirs, completeRustFiles)
+
+        val ignoreGlobs = patternType == WORKSPACE &&
+            element.parentOfType<TomlKeyValue>()?.key?.segments?.firstOrNull()?.name in KEYS_SUPPORTING_GLOB
+        return if (ignoreGlobs) {
+            referenceSet.referencesUntilFirstGlob.toList().toTypedArray()
+        } else {
+            referenceSet.allReferences
+        }
```

Comment:
Maybe it's better just to pass a flag to `CargoTomlFileReferenceSet` not to produce references for glob patterns?
I suppose it should simplify implementation

Reply:
Ok, although then we need to introduce a new getter for references anyway, since allReferences cannot be used directly (because of the problem with nulls returned from createFileReference).

Reply:
> because of the problem with nulls returned from createFileReference

What problem are you talking about?

Reply:
Sorry for the confusion, I meant this: https://github.com/intellij-rust/intellij-rust/pull/7403#discussion_r659176699.
It seemed to me that it would be fragile to filter the references directly in `createFileReference`, because it would need to become stateful (it would need to return `null` after the first glob reference).

But since now `CargoTomlFileReferenceSet` is stateful anyway, it probably doesn't matter. Would this be ok?

```kotlin
// ignoreGlobReferences is a flag passed to CargoTomlFileReferenceSet

override fun createFileReference(range: TextRange, index: Int, text: String): FileReference? {
        val reference = super.createFileReference(range, index, text) ?: return null
        if (ignoreGlobReferences) {
            if (!globFound && isGlobPathFragment(text)) {
                globFound = true
            }
            if (globFound) return null
        }
        return reference
    }
```

Reply:
Yep, looks ok for me and it should work as expected

Reply:
Ah, dammit. Virtual calls in constructors strike back again :(
`FileReferenceSet` calls reparse in its constructor, therefore the `ignoreGlobReferences` is not set properly when `reparse` is executed :(

We could call reparse manually after the object is fully constructed, but that is a nasty hack.
So I guess let's introduce additional property that will filter the references if the flag is set?

Reply:
You can create a child class of `CargoTomlFileReferenceSet` where you can filter glob references unconditionally.
And choose which class you should use in `getReferencesByElement` to avoid this problem

Reply:
Interesting fact: there is another constructor `com.intellij.psi.impl.source.resolve.reference.impl.providers.FileReferenceSet#FileReferenceSet(java.lang.String, com.intellij.psi.PsiElement, int, com.intellij.psi.PsiReferenceProvider, boolean, boolean, com.intellij.openapi.fileTypes.FileType[], boolean)` where you can pass `init` flag to control if `reparse` call is needed in  super constructor

Reply:
Hmm, using a child class is probably the cleanest solution. I changed it :)